tests/smoke-tests/test-dynamic-fees.ts

```diff
@@ -0,0 +1,241 @@
+      failures.forEach(({ blockNum, fillPermill, change }) => {
+        debug(
+          `Block #${blockNum} is ${(fillPermill.toNumber() / 1_000_000).toFixed(
+            2
+          )}% full with feeMultiplier ${change}`
+        );
+      });
+
+      expect(
+        failures.length,
+        `Please investigate blocks ${failures.map(({ blockNum }) => blockNum).join(`, `)}`
+      ).to.equals(0);
+    });
+
+    testIt("C200", "Block utilization by gas corresponds to fee multiplier", async function () {
+      const enriched = blockData.map(({ blockNum, ethersBlock, nextFeeMultiplier }) => {
+        const fillPermill = ethersBlock.gasUsed.mul("1000000").div(ethersBlock.gasLimit);
+        const change = checkMultiplier(
+          blockData.find((a) => a.blockNum == blockNum - 1),
+          nextFeeMultiplier
+        );
+        const valid = isChangeDirectionValid(
+          new BN(fillPermill.toString()),
+          change,
+          nextFeeMultiplier
+        );
+
+        return { blockNum, fillPermill, change, valid };
+      });
+
+      const failures = enriched.filter(({ valid }) => !valid);
+      failures.forEach(({ blockNum, fillPermill, change }) => {
+        debug(
+          `Block #${blockNum} is ${(fillPermill.toNumber() / 1_000_000).toFixed(
+            2
+          )}% full with feeMultiplier ${change}`
+        );
+      });
+      expect(
+        failures.length,
+        `Please investigate blocks ${failures.map(({ blockNum }) => blockNum).join(`, `)}`
+      ).to.equals(0);
+    });
+
+    testIt("C300", "BaseFeePerGas is within expected min/max", function () {
+      const failures = blockData.filter(({ baseFeePerGasInGwei }) => {
+        return (
+          ethers.utils
+            .parseUnits(baseFeePerGasInGwei, "gwei")
+            .lt(BigNumber.from(RUNTIME_CONSTANTS[runtime].MIN_BASE_FEE_IN_WEI)) ||
+          ethers.utils
+            .parseUnits(baseFeePerGasInGwei, "gwei")
+            .gt(BigNumber.from(RUNTIME_CONSTANTS[runtime].MAX_BASE_FEE_IN_WEI))
+        );
+      });
+
+      failures.forEach(({ blockNum, baseFeePerGasInGwei }) => {
+        debug(`Block #${blockNum} has baseFeePerGas out of range: ${baseFeePerGasInGwei}`);
+      });
+      expect(
+        failures.length,
+        `Please investigate blocks ${failures.map(({ blockNum }) => blockNum).join(`, `)}`
+      ).to.equals(0);
+    });
+
+    testIt("C400", "BaseFeePerGas is correctly calculated", function () {
+      const supplyFactor =
+        context.polkadotApi.consts.system.version.specName.toString() === "moonbeam" ? 100n : 1n;
+
+      const failures = blockData
+        .map(({ blockNum, nextFeeMultiplier, baseFeePerGasInGwei }) => {
+          const baseFeePerGasInWei = ethers.utils.parseUnits(baseFeePerGasInGwei, "gwei");
+
+          const expectedBaseFeePerGasInWei =
+            (nextFeeMultiplier.toBigInt() * WEIGHT_FEE * WEIGHT_PER_GAS * supplyFactor) /
+            ethers.utils.parseEther("1").toBigInt();
```

Comment:
This is the lossy part of the equation, as expected. What could be useful is to show that the loss (the truncation) here is always less than some amount.

Reply:
...or maybe not, that might just look very redundant...

Reply:
Have added a new test which checks that fee charged = gas used * weight_to_gas const.

I think this isn't what you were exactly asking for, but by inference we have enough coverage of the individual components that fee->gas has been derisked for it to be low concern.